### ISSUE TITLE
Support for pull request webhook events

### ISSUE THREAD

**Issue by sandro-h** [gocd/gocd#6668]
##### Issue Type

- Feature proposal

##### Summary

Support webhooks to trigger pipelines when a pull request is created or updated; based on the [gocd-build-github-pull-requests plugin](https://github.com/ashwanthkumar/gocd-build-github-pull-requests).

##### Motivation

My understanding is that the GoCD team favors trunk-based development. That said, there is already above plugin to support PR pipelines to build changes in a pull request and based on feedback, quite a few users of GoCD do want to do pull-request based development.

One of the main problems we encounter at our company is that the plugin is forced to combine changes in multiple PRs into a single pipeline execution if they all happen relatively close to each other. As I understand, this is an issue that cannot be fixed in the plugin because that is how GoCD handles material updates (see https://github.com/gocd/gocd/issues/938).

One mitigating approach we found is to trigger the material update as soon as a change in a PR happens instead of waiting for the polling interval. And the most appropriate way to do this is with webhooks triggered on new/updated pull-requests in the SCM tool.
Even if the above problem could be solved by adapting the material update strategy of GoCD, it would still be useful to have webhooks to trigger them for faster response times and to decrease the polling load on the SCM tool.

##### Proposal

GoCD already offers webhook endpoints for commits on a repository for various SCM tools (github, gitlab, bitbucket). These could be extended fairly easily to also accept pull-request events. These events would have to trigger an update of the corresponding pluggable scm materials which in turn trigger the PR pipelines.
I have PoC'd this for Bitbucket Server Edition because that happens to be what we use at our company. The change can be seen here: https://github.com/adnovum/gocd/commit/9957c6b4fe83c7e59d6354e09ea0e9eccf826299
The basic outline is:
- Accept pr events in the webhook endpoint
- Extract the git repo url from the event
- Check all pluggable scm materials (that have autoUpdate disabled) for whether they have a "url" config parameter and compare this to the url from the event.

There are a number of things to clarify before something like that could be used and that's where I'm interested in your feedback:
- First of all, is the proposed feature even something you consider reasonable and useful?
- How should this be generalized. Specifically what pluggable scm materials should be used when looking for materials to update in a PR webhook. Should this be configurable? What material config property (currently "url") should be used to find the right material.
- Related to the previous question, should/could this be generalized to trigger arbitrary (pluggable) materials for the normal commit webhook event, instead of the git material only. The reason I ask is that BitBucket server edition doesn't actually have a pr:updated event for commit changes in the PR, so this could be mitigated by triggering the pr material on commits in the branch used by the PR. Sadly it doesn't look like the BB team will change that any time soon. But I understand if that is too specific of a use case for you.
- Assuming the other questions can be answered in a satisfying way, would you accept a PR to enhance just the bitbucket server endpoint, or would you require all other current scm tools to also be supported immediately. That's possible of course, I just haven't looked at all the others and it would take some more time to get there.

Thanks for taking this into consideration!

**Comment by maheshp** [comment]
Hello @sandro-h 

Apologies for the delayed response, my thoughts below,

> First of all, is the proposed feature even something you consider reasonable and useful?

Having a webhook which accepts PR events will be useful and would solve some of the problems mentioned by you around faster pipeline scheduling and reduced polling load. 

But, I am afraid it may not completely solve your original problem. Even after using webhooks there could be cases where a pipeline can be triggered with changeset from multiple PRs. Consider a scenario where the first stage of the pipeline is scheduled and building, meanwhile if there are multiple webhook events this will trigger multiple material updates and the subsequent run of the pipeline would be against multiple PRs.

> How should this be generalized. Specifically what pluggable scm materials should be used when looking for materials to update in a PR webhook. Should this be configurable? What material config property (currently "url") should be used to find the right material.

Relying on the `url` to identify the material seems reasonable. Currently I cannot think of any other way to identify a pluggable material apart from the `url`.  I don't see a need to have the materials configurable.

> Related to the previous question, should/could this be generalized to trigger arbitrary (pluggable) materials for the normal commit webhook event, instead of the git material only. The reason I ask is that BitBucket server edition doesn't actually have a pr:updated event for commit changes in the PR, so this could be mitigated by triggering the pr material on commits in the branch used by the PR. Sadly it doesn't look like the BB team will change that any time soon. But I understand if that is too specific of a use case for you.

If I understand right you are looking at a commit triggering either or both a git material as well the pluggable scm material if the url matches. This seems ok, but I need to think a bit more and get back to you.

> Assuming the other questions can be answered in a satisfying way, would you accept a PR to enhance just the bitbucket server endpoint, or would you require all other current scm tools to also be supported immediately. That's possible of course, I just haven't looked at all the others and it would take some more time to get there.

A PR is always welcome, for now we can start with supporting bitbucket and later add it other scm tools.

**Comment by sandro-h** [comment]
Hi @maheshp,

Sorry for the delayed answer, I was on vacation. Thanks for the encouraging response.

> But, I am afraid it may not completely solve your original problem. Even after using webhooks there could be cases where a pipeline can be triggered with changeset from multiple PRs. Consider a scenario where the first stage of the pipeline is scheduled and building, meanwhile if there are multiple webhook events this will trigger multiple material updates and the subsequent run of the pipeline would be against multiple PRs.

True, I was maybe under the false impression that multiple _individual_ material updates triggered by a webhook would each _queue_ up a pipeline run (as opposed to one material update containing multiple changes triggering a single pipeline run). So if GoCD receives multiple updates from a material while the pipeline is already running, it will conflate them into a single pipeline run as well, correct?
I suppose that fits a related problem we also encounter: that it's currently not possible to run the same stage of a pipeline in parallel executions, thus blocking additional runs.
So yes, the webhooks probably wouldn't solve all our problems, but at least some of them and either way it would be useful to have more responsive PR builds by using webhooks. Hopefully the related problems can be solved eventually.

> Relying on the url to identify the material seems reasonable. Currently I cannot think of any other way to identify a pluggable material apart from the url. I don't see a need to have the materials configurable.

Great, then I will keep the simple approach of checking for a `url` parameter.

> If I understand right you are looking at a commit triggering either or both a git material as well the pluggable scm material if the url matches. This seems ok, but I need to think a bit more and get back to you.

Yes, exactly.

> A PR is always welcome, for now we can start with supporting bitbucket and later add it other scm tools.

Thanks, that makes the PR scope a bit more manageable for me. I will flesh out the current solution a bit more with your feedback in mind.

**Comment by sandro-h** [comment]
> True, I was maybe under the false impression that multiple individual material updates triggered by a webhook would each queue up a pipeline run (as opposed to one material update containing multiple changes triggering a single pipeline run).

I just now realized that my understanding was wrong from the beginning. I had thought that the gocd-build-github-pull-requests plugin will send a list of all changes across multiple PRs in one response, and this list is what gocd conflates. But the plugin actually already throttles itself and only sends one change back per polling response. So what GoCD actually conflates is the multiple responses from the plugin that might accrue while the pipeline is running. That matches what you pointed out.

**Comment by maheshp** [comment]
> So if GoCD receives multiple updates from a material while the pipeline is already running, it will conflate them into a single pipeline run as well, correct?

Thats right.

> I just now realized that my understanding was wrong from the beginning. I had thought that the gocd-build-github-pull-requests plugin will send a list of all changes across multiple PRs in one response, and this list is what gocd conflates. But the plugin actually already throttles itself and only sends one change back per polling response. So what GoCD actually conflates is the multiple responses from the plugin that might accrue while the pipeline is running. That matches what you pointed out.

Yeah, the plugin behaviour mentioned by you is right. Since GoCD combines changes all changes it might look at the pipeline is run against all the PR's but actually it would just run against a specific commit in a PR. E.g check [this](https://build.gocd.org/go/pipelines/trigger/4979/do-nothing/1/materials) pipeline run on build.gocd.org, the changes listed are across multiple PR's but the pipeline run is against the first commit(7da84f3b8cebff28a9a579b6a6c59f420a246341) in the list.

**Comment by lukyanov** [comment]
There is one thing I don't understand about GoCD related to the subject. Can the pipeline be run multiple times in parallel at the same time? I thought it could. Not from the UI though, but by triggering changes in materials while the pipeline is already running. Is this true?

If so, the PR webhook proposed here would help, wouldn't it? While the pipeline is running, there is another set of changes in the material waiting for the pipeline execution and the pipeline will run without waiting for the previous run to finish, won't it?

If this is not true, then what the checkbox "Run single instance of pipeline at a time " in the settings mean?

**Comment by lukyanov** [comment]
I think I found some information about this. It looks like GoCD can't run multiple instances of the same stage. So if the first stage is in the progress, GoCD will be collecting changes in the material and will not trigger the second instance, right?

I have an idea then. What if we add a very short first stage (say with an empty task)? That would mean while the pipeline is executing the second (long-running) stage, GoCD will still be able to trigger the pipeline from the beginning for another PR?

Does it make sense?

P.S. I know it's all about workarounds. But I really don't know how to live without PRs.

**Comment by maheshp** [comment]
> So if the first stage is in the progress, GoCD will be collecting changes in the material and will not trigger the second instance, right?

Yes, you are right. GoCD does not allow parallel runs of a stage. If the stage in question is the first stage of the pipeline, GoCD would not schedule a new instance of the pipeline if the first stage is in progress.

As you mentioned having a short stage is kind of a workaround but it will still not guarantee a pipeline run against each PR.
